src/Features/Core/Portable/Completion/Providers/AbstractTypeImportCompletionProvider.cs

```diff
@@ -0,0 +1,316 @@
+                            declarationsInReference = _typeImportCompletionService.GetAccessibleTopLevelTypesFromPEReference(
+                                project.Solution,
+                                compilation,
+                                peReference,
+                                namespacesInScope,
+                                cancellationToken);
+                        }
+
+                        builder.AddRange(declarationsInReference);
+                    }
+
+                    return ImmutableArray<CompletionItem>.CastUp(builder.ToImmutableAndFree());
+                }
+            }
+
+            return ImmutableArray<CompletionItem>.Empty;
+        }
+
+        public override async Task<CompletionChange> GetChangeAsync(Document document, CompletionItem item, char? commitKey = default, CancellationToken cancellationToken = default)
+        {
+            var newDocument = await ComputeNewDocumentAsync(document, item, cancellationToken).ConfigureAwait(false);
+            var newText = await newDocument.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var newRoot = await newDocument.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            int? newPosition = null;
+
+            // Attempt to find the inserted node and move the caret appropriately
+            if (newRoot != null)
+            {
+                var caretTarget = newRoot.GetAnnotatedNodesAndTokens(_annotation).FirstOrNullable();
+                if (caretTarget != null)
+                {
+                    var targetPosition = caretTarget.Value.AsNode().GetLastToken().Span.End;
+
+                    // Something weird happened and we failed to get a valid position.
+                    // Bail on moving the caret.
+                    if (targetPosition > 0 && targetPosition <= newText.Length)
+                    {
+                        newPosition = targetPosition;
+                    }
+                }
+            }
+
+            var changes = await newDocument.GetTextChangesAsync(document, cancellationToken).ConfigureAwait(false);
+            var change = Utilities.Collapse(newText, changes.ToImmutableArray());
+
+            return CompletionChange.Create(change, newPosition, includesCommitCharacter: true);
+        }
+
+        private async Task<Document> ComputeNewDocumentAsync(Document document, CompletionItem completionItem, CancellationToken cancellationToken)
+        {
+            var tree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            var root = await tree.GetRootAsync(cancellationToken).ConfigureAwait(false);
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
+            // Complete type name.
+            var finalText = root.GetText(text.Encoding)
+                .Replace(completionItem.Span, completionItem.DisplayText);
+            document = document.WithText(finalText);
```

Comment:
replacing document, then havin to go re-get the data off of it is pretty confsing.  can you use anotehr name here, or pass to a helper method?  that will keep things nice and segmented.

Reply:
Fixed